sklearn/metrics/_pairwise_distances_reduction.pyx

```diff
@@ -629,15 +664,16 @@ cdef class PairwiseDistancesArgKmin(PairwiseDistancesReduction):
 
         return pda._finalize_results(return_distance)
 
-    def __cinit__(
+    def __init__(
```

Comment:
Based on my understanding of `__cinit__`, we need to do something like this:

```python
cdef class PairwiseDistancesReduction:
    def __cinit__(...):
        self.heaps_r_distances_chunks = <DTYPE_t **> malloc(
            sizeof(DTYPE_t *) * self.chunks_n_threads
        )
        ...
```

and in the subclass `FastEuclideanPairwiseDistancesArgKmin`:

```python
cdef class FastEuclideanPairwiseDistancesArgKmin:
    def __cinit__(...):
        # Do not need to call `super()` here. Cython handles it for you
        self.dist_middle_terms_chunks = <DTYPE_t **> malloc(
             sizeof(DTYPE_t *) * self.chunks_n_threads
         )
```

Based on [cython docs](https://cython.readthedocs.io/en/latest/src/userguide/special_methods.html#initialisation-methods-cinit-and-init):

> If your extension type has a base type, any existing __cinit__() methods in the base type hierarchy are automatically called before your __cinit__() method.

Reply:
Yes, it's just that `__init__` has to have the same signature and arguments than `__cinit__`.

It might be possible via using args and kwargs, but might also make the signature more complex. I am currently thinking about it.

Reply:
03b6c79 resolves this.

Reply:
Seeing the diff to make `__cinit__` work, I am -1 on the idea because of how much more complexity it adds.

Do you think there are the consequences for using `__init__` for `malloc`?

Reply:
> Seeing the diff to make __cinit__ work, I am -1 on the idea because of how much more complexity it adds.

So am I.

> Do you think there are any consequences for using __init__ for malloc?

`__cinit__` is always called. There's no such guarantee for `__init__`.
In some cases (e.g. for de-serialization or when creating an instance via `__new__`), `__init__` won't be called.

If those interfaces were to be public, such problem would appear (because data-structures would not be allocated on deserialization in the user code).
 
In our case, I think it is OK-ish, because:
 - they are allocated in Python via `__init__` in a private implementation. 
 - those aren't attributes of estimators hence we do not have the problem with deserialization.
 
I would be in favor of reverting everything to use `__init__` and add a comment motivating this choice.

What do you think?

Reply:
I reverted those changes to solely use `__init__` in the last commits.

Let me know if this works for you.

Reply:
> Let me know if this works for you.

It works for me.